### PR TITLE
Adjust polygon position heights before removing duplicates if not perPositionHeight

### PR DESCRIPTION
**Pull request by hpinkos**
Fixes crash reported in https://github.com/AnalyticalGraphicsInc/cesium/issues/4863#issuecomment-399863982

If `perPositionHeight: false`, the positions passed to `PolygonGeometry` are scaled to the ellipsoid surface then adjusted for `height` and `extrudedHeight`.  However, this needs to happen before `arrayRemoveDuplicates` for the case where there are positions with the same lon/lat but different heights.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Signed CLA is on file.

@hpinkos, thanks for the pull request! Maintainers, we have a signed CLA from @hpinkos, so you can review this at any time.

:warning: I noticed that [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md) has not been updated. If this change updates the public API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and comment on this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by mramato** [comment]
@likangning93 can you review.

Please update CHANGES.

File: Source/Core/PolygonGeometryLibrary.js [hunk 1]
```diff
@@ -217,6 +217,13 @@ define([
             var outerRing = outerNode.positions;
             var holes = outerNode.holes;
 
+            var i;
+            if (!perPositionHeight) {
+                for (i = 0; i < outerRing.length; i++) {
```
Comment (likangning93) [inline Source/Core/PolygonGeometryLibrary.js]: Should this get set on a temp var before the loop?

File: Source/Core/PolygonGeometryLibrary.js [hunk 2]
```diff
@@ -234,12 +241,18 @@ define([
...
             var polygonHoles = [];
-            var i;
             var j;
 
             for (i = 0; i < numChildren; i++) {
                 var hole = holes[i];
-                var holePositions = arrayRemoveDuplicates(hole.positions, Cartesian3.equalsEpsilon, true);
+                var holePositions = hole.positions;
+                if (!perPositionHeight) {
+                    for (j = 0; j < holePositions.length; ++j) {
```
Comment (likangning93) [inline Source/Core/PolygonGeometryLibrary.js]: Same as above.

File: Source/Core/PolygonOutlineGeometry.js [hunk 3]
```diff
@@ -483,9 +483,15 @@ define([
         var queue = new Queue();
         queue.enqueue(polygonHierarchy);
         var i;
+        var j;
         while (queue.length !== 0) {
             var outerNode = queue.dequeue();
             var outerRing = outerNode.positions;
+            if (!perPositionHeight) {
+                for (i = 0; i < outerRing.length; i++) {
```
Comment (likangning93) [inline Source/Core/PolygonOutlineGeometry.js]: Same as above.

File: Source/Core/PolygonOutlineGeometry.js [hunk 4]
```diff
@@ -495,18 +501,24 @@ define([
             // The outer polygon contains inner polygons
             for (i = 0; i < numChildren; i++) {
                 var hole = outerNode.holes[i];
-                hole.positions = arrayRemoveDuplicates(hole.positions, Cartesian3.equalsEpsilon, true);
-                if (hole.positions.length < 3) {
+                var holePositions = hole.positions;
+                if (!perPositionHeight) {
+                    for (j = 0; j < holePositions.length; ++j) {
```
Comment (likangning93) [inline Source/Core/PolygonOutlineGeometry.js]: Same as above.

File: Specs/Core/PolygonOutlineGeometrySpec.js [hunk 5]
```diff
@@ -96,6 +96,46 @@ defineSuite([
...
+                positions : Cartesian3.fromDegreesArrayHeights([
+                    0.0, 0.0, 10.0,
+                    0.0, 0.0, 20.0,
+                    0.0, 0.0, 30.0
+                ])
+            }]
+        };
+
+        var geometry = PolygonOutlineGeometry.createGeometry(new PolygonOutlineGeometry({ polygonHierarchy : hierarchy, perPositionHeight: true }));
+        expect(geometry).not.toBeUndefined();
```
Comment (likangning93) [inline Specs/Core/PolygonOutlineGeometrySpec.js]: `.toBeDefined()`

File: Specs/Core/PolygonGeometrySpec.js [hunk 6]
```diff
@@ -84,24 +84,64 @@ defineSuite([
...
+                positions : Cartesian3.fromDegreesArrayHeights([
+                    0.0, 0.0, 10.0,
+                    0.0, 0.0, 20.0,
+                    0.0, 0.0, 30.0
+                ])
+            }]
+        };
+
+        var geometry = PolygonGeometry.createGeometry(new PolygonGeometry({ polygonHierarchy : hierarchy, perPositionHeight: true }));
+        expect(geometry).not.toBeUndefined();
```
Comment (likangning93) [inline Specs/Core/PolygonGeometrySpec.js]: `.toBeDefined()`

**Comment by hpinkos** [comment]
Thanks @likangning93! Updated

**Pull request merged**